### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by harryliu-intel**
A few performance improvements:
- replace probe() plus busy waiting with select()
- use a specialized function to compare messages of enum constants
- use multiprocessing semaphore to avoid overhead of pipe

### CODE REVIEW

**Comment by awintel** [comment]
Oh, oh... that means work for me. Not knowing what you would be doing. I have also refactored the LoihiPyProcessModel by pushing a lot of the generic MGMT command handling into the AbstractPyProcModel since this is not Loihi specific and should also be the same for AsyncProcs e.g. on a CPU.

For that purpose I introduced a message handler based on a dictionary mapping from commands to function objects/message handlers. The AbstractPyProcModel sets up a basic dictionary with RUN, PAUSE, STOP and respective message handlers. Sub classes like LoihiPyProcModel would then just extend the set of handlers. This made the code much cleaner.

Will first finish my stuff then revisit your changes.

What in all of your changes was the key to making it more performant and by how much?

**Comment by mgkwill** [comment]
I like the enum_equal fucntion and using multiprocessing semaphore also seems like a good idea. 

This change also has the likelihood of affect current inflight Pyport changes but will adapt as necessary.

**Comment by harryliu-intel** [comment]
Sorry, Andreas, I didn't know you'd be working in the same area.  The most significant speed up comes from the select() commit, but I think each of the other 2 commits also makes a measurable difference.

**Comment by awintel** [comment]
No worries. Not your fault ;-)

File: src/lava/magma/core/model/py/model.py [hunk 1]
```diff
@@ -148,37 +147,44 @@ def run(self):
...
+
+            channel_actions = [(self.service_to_process_cmd, lambda: 'cmd')]
+            if enum_equal(phase, PyLoihiProcessModel.Phase.PRE_MGMT) or \
+                    enum_equal(phase, PyLoihiProcessModel.Phase.POST_MGMT):
+                for var_port in self.var_ports:
+                    for csp_port in var_port.csp_ports:
+                        if isinstance(csp_port, CspRecvPort):
+                            channel_actions.append((csp_port, lambda: var_port))
+            elif enum_equal(phase, PyLoihiProcessModel.Phase.HOST):
+                channel_actions.append((self.service_to_process_req,
```
Comment (awintel) [inline src/lava/magma/core/model/py/model.py:170]: From a type perspective it is a bit inconsistent to give either a string or a port back as an action.

File: src/lava/magma/compiler/channels/pypychannel.py [hunk 2]
```diff
@@ -265,6 +273,43 @@ def join(self):
...
+class CspSelector:
+    """
+    Utility class to allow waiting for multiple channels to become ready
+    """
+
+    def __init__(self):
+        """Instantiates CspSelector object and class attributes"""
+        self._cv = Condition()
+
+    def _changed(self, channel):
```
Comment (awintel) [inline src/lava/magma/compiler/channels/pypychannel.py:285]: channel is not used

File: src/lava/magma/compiler/channels/pypychannel.py [hunk 3]
```diff
@@ -97,7 +98,10 @@ def _ack_callback(self):
         try:
             while not self._done:
                 self._ack.recv_bytes(0)
+                not_full = self.probe()
                 self._semaphore.release()
+                if self.observer and not not_full:
+                    self.observer(self)
```
Comment (awintel) [inline src/lava/magma/compiler/channels/pypychannel.py:104]: CspSelector._changed(..) does not need its input argument (self in this case).

File: src/lava/magma/compiler/channels/pypychannel.py [hunk 4]
```diff
@@ -265,6 +273,43 @@ def join(self):
...
+        *args: ty.Tuple[ty.Union[CspSendPort, CspRecvPort],
+                        ty.Callable[[], ty.Any]
+                        ]
+    ):
+        """
+        Wait for any channel to become ready, then execute the corresponding
+        callable and return the result.
+        """
+        with self._cv:
+            self._set_observer(args, self._changed)
```
Comment (awintel) [inline src/lava/magma/compiler/channels/pypychannel.py:304]: Why do we have to set and unset this all the time?

File: src/lava/magma/compiler/channels/pypychannel.py [hunk 5]
```diff
@@ -230,7 +235,10 @@ def _req_callback(self):
         try:
             while not self._done:
                 self._req.recv_bytes(0)
+                not_empty = self.probe()
```
Comment (awintel) [inline src/lava/magma/compiler/channels/pypychannel.py:238]: This loop is still busy waiting, isn't it? Is this a performance concern? Could this also be improved in performance by using a condition?

**Review by awintel — COMMENTED** [review]
Took me a bit to understand what's happening here...

My interpretation is the following:
- You introduced a CspSelector which essentially has a threading.Condition object. 
- CspSelector.select is like a message handler of its own. It is given a list of channels and actions to perform when there is a token on any of the given channels.
- The select method passes the central Condition of the CspSelector to each channel.
- The CspSelector then simply waits for any channel to trigger the condition. This Condition.wait() is probably implemented more efficiently than busy waiting.
- Once any channel has a new token it sends a notification via the threading.Condition to all observers. This unblocks the wait() inside the CspSelector.select() method. We iterate once over all all channels to see which has as token and then call the corresponding action/message handler.

Overall this looks good!
I don't mind if we pull this into the current release if this all works but going forward I hope we can simplify the overall design of the PyRuntimeService and PyProcessModel as outlined in the other git issue:
https://github.com/lava-nc/lava/issues/86

If possible, we should treat all channels equally, have one selector that listens on all available channels and calls an appropriate action when any becomes active. The action could just be a function pointer instead of a string in some cases or  port in some other case.

File: src/lava/magma/compiler/channels/pypychannel.py (hunk 4 above)
Comment (harryliu-intel) [reply src/lava/magma/compiler/channels/pypychannel.py:304]: In an earlier version, all channels notified, and _changed() determined if the source channel is part of the current set being selected on.  (This is also the reason for the now unused channel argument, which I didn't bother to remove.)  I thought doing it this way was cheaper and cleaner.

File: src/lava/magma/compiler/channels/pypychannel.py (hunk 5 above)
Comment (harryliu-intel) [reply src/lava/magma/compiler/channels/pypychannel.py:238]: There no busy waiting here, because self._req.recv_bytes() would have blocked.

**Review by harryliu-intel — COMMENTED** [review]
> The action could just be a function pointer instead of a string in some cases or port in some
> other case.

I agree but I avoided that because it would have required larger structural changes that might make this commit more difficult to understand and review.  I would suggest the original authors come back and refactor it.

**Review by PhilippPlank — APPROVED** [review]
Overall I think we should merge it, since it improves performance and does not break our unit tests. We still need to refactor our runtime/runtime_service interactions, but one step after the other.

**Review by mgkwill — APPROVED** [review]

**Pull request merged**